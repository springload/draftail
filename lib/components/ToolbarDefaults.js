// @flow
import React, { PureComponent } from "react";

import ToolbarButton from "./ToolbarButton";
import ToolbarGroup from "./ToolbarGroup";

import {
  BR_TYPE,
  ENTITY_TYPE,
  UNDO_TYPE,
  REDO_TYPE,
  LABELS,
  DESCRIPTIONS,
} from "../api/constants";
import behavior from "../api/behavior";
import type { IconProp } from "./Icon";

type ControlProp = {
  // Describes the control in the editor UI, concisely.
  label?: string,
  // Describes the control in the editor UI.
  description?: string,
  // Represents the control in the editor UI.
  icon?: IconProp,
};

const getButtonLabel = (type: string, config: boolean | ControlProp) => {
  const icon = typeof config === "boolean" ? undefined : config.icon;

  if (typeof config.label === "string") {
    return config.label;
  }

  if (typeof icon !== "undefined") {
    return null;
  }

  return LABELS[type];
};

const getButtonTitle = (type: string, config: boolean | ControlProp) => {
  const description =
    typeof config === "boolean" || typeof config.description === "undefined"
      ? DESCRIPTIONS[type]
      : config.description;
  const hasShortcut = behavior.hasKeyboardShortcut(type);
  let title = description;

  if (hasShortcut) {
    const desc = description ? `${description}\n` : "";
    title = `${desc}${behavior.getKeyboardShortcut(type)}`;
  }

  return title;
};

export type ToolbarDefaultProps = {
  currentStyles: {
    has: (style: string) => boolean,
  },
  currentBlock: string,
  enableHorizontalRule: boolean | ControlProp,
  enableLineBreak: boolean | ControlProp,
  showUndoControl: boolean | ControlProp,
  showRedoControl: boolean | ControlProp,
  entityTypes: $ReadOnlyArray<{
    ...ControlProp,
    type: string,
  }>,
  blockTypes: $ReadOnlyArray<{
    ...ControlProp,
    type: string,
  }>,
  inlineStyles: $ReadOnlyArray<{
    ...ControlProp,
    type: string,
  }>,
  toggleBlockType: (blockType: string) => void,
  toggleInlineStyle: (inlineStyle: string) => void,
  addHR: () => void,
  addBR: () => void,
  onUndoRedo: (type: string) => void,
  onRequestSource: (entityType: string) => void,
};

class ToolbarDefaults extends PureComponent<ToolbarDefaultProps> {
  render() {
    const {
      currentStyles,
      currentBlock,
      blockTypes,
      inlineStyles,
      enableHorizontalRule,
      enableLineBreak,
      showUndoControl,
      showRedoControl,
      entityTypes,
      toggleBlockType,
      toggleInlineStyle,
      addHR,
      addBR,
      onUndoRedo,
      onRequestSource,
    } = this.props;
    return [
      <ToolbarGroup key="styles">
        {inlineStyles.map((t) => (
          <ToolbarButton
            key={t.type}
            name={t.type}
            active={currentStyles.has(t.type)}
            label={getButtonLabel(t.type, t)}
            title={getButtonTitle(t.type, t)}
            icon={t.icon}
            onClick={toggleInlineStyle}
          />
        ))}
      </ToolbarGroup>,

      <ToolbarGroup key="blocks">
        {blockTypes.map((t) => (
          <ToolbarButton
            key={t.type}
            name={t.type}
            active={currentBlock === t.type}
            label={getButtonLabel(t.type, t)}
            title={getButtonTitle(t.type, t)}
            icon={t.icon}
            onClick={toggleBlockType}
          />
        ))}
      </ToolbarGroup>,

      <ToolbarGroup key="hr-br">
        {enableHorizontalRule ? (
          <ToolbarButton
            name={ENTITY_TYPE.HORIZONTAL_RULE}
            onClick={addHR}
            label={getButtonLabel(
              ENTITY_TYPE.HORIZONTAL_RULE,
              enableHorizontalRule,
            )}
            title={getButtonTitle(
              ENTITY_TYPE.HORIZONTAL_RULE,
              enableHorizontalRule,
            )}
            icon={
              typeof enableHorizontalRule !== "boolean"
                ? enableHorizontalRule.icon
                : null
            }
          />
        ) : null}

        {enableLineBreak ? (
          <ToolbarButton
            name={BR_TYPE}
            onClick={addBR}
            label={getButtonLabel(BR_TYPE, enableLineBreak)}
            title={getButtonTitle(BR_TYPE, enableLineBreak)}
            icon={
              typeof enableLineBreak !== "boolean" ? enableLineBreak.icon : null
            }
          />
        ) : null}
      </ToolbarGroup>,

      <ToolbarGroup key="entities">
        {entityTypes.map((t) => (
          <ToolbarButton
            key={t.type}
            name={t.type}
            onClick={onRequestSource}
            label={getButtonLabel(t.type, t)}
            title={getButtonTitle(t.type, t)}
            icon={t.icon}
          />
        ))}
      </ToolbarGroup>,

      <ToolbarGroup key="undo-redo">
        {showUndoControl ? (
          <ToolbarButton
            name={UNDO_TYPE}
            onClick={onUndoRedo}
            label={getButtonLabel(UNDO_TYPE, showUndoControl)}
            title={getButtonTitle(UNDO_TYPE, showUndoControl)}
            icon={
              typeof showUndoControl !== "boolean" ? showUndoControl.icon : null
            }
          />
        ) : null}

        {showRedoControl ? (
          <ToolbarButton
            name={REDO_TYPE}
            onClick={onUndoRedo}
            label={getButtonLabel(REDO_TYPE, showRedoControl)}
            title={getButtonTitle(REDO_TYPE, showRedoControl)}
            icon={
              typeof showRedoControl !== "boolean" ? showRedoControl.icon : null
            }
          />
        ) : null}
      </ToolbarGroup>,
    ];
  }
}

export default ToolbarDefaults;
