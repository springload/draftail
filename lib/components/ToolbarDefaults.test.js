import { OrderedSet } from "immutable";
import React from "react";
import { shallow, mount } from "enzyme";

import ToolbarDefaults from "./ToolbarDefaults";

const mockProps = {
  currentStyles: new OrderedSet(),
  currentBlock: "unstyled",
  enableHorizontalRule: false,
  enableLineBreak: false,
  showUndoControl: false,
  showRedoControl: false,
  entityTypes: [],
  blockTypes: [],
  inlineStyles: [],
  toggleBlockType: () => {},
  toggleInlineStyle: () => {},
  addHR: () => {},
  addBR: () => {},
  onUndoRedo: () => {},
  onRequestSource: () => {},
};

describe("ToolbarDefaults", () => {
  it("empty", () => {
    expect(shallow(<ToolbarDefaults {...mockProps} />)).toMatchSnapshot();
  });

  describe("#enableHorizontalRule", () => {
    it("works", () => {
      expect(
        mount(<ToolbarDefaults {...mockProps} enableHorizontalRule />),
      ).toMatchSnapshot();
    });

    it("control overrides", () => {
      expect(
        mount(
          <ToolbarDefaults
            {...mockProps}
            enableHorizontalRule={{
              icon: "#icon-hr",
              label: "HR",
              description: "Horizontal rule",
            }}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  describe("#enableLineBreak", () => {
    it("works", () => {
      expect(
        mount(<ToolbarDefaults {...mockProps} enableLineBreak />),
      ).toMatchSnapshot();
    });

    it("control overrides", () => {
      expect(
        mount(
          <ToolbarDefaults
            {...mockProps}
            enableLineBreak={{
              icon: "#icon-br",
              label: "BR",
              description: "Soft line break",
            }}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  describe("#showUndoControl", () => {
    it("works", () => {
      expect(
        mount(<ToolbarDefaults {...mockProps} showUndoControl />),
      ).toMatchSnapshot();
    });

    it("control overrides", () => {
      expect(
        mount(
          <ToolbarDefaults
            {...mockProps}
            showUndoControl={{
              icon: "#icon-undo",
              label: "Undo",
              description: "Undo last change",
            }}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  describe("#showRedoControl", () => {
    it("works", () => {
      expect(
        mount(<ToolbarDefaults {...mockProps} showRedoControl />),
      ).toMatchSnapshot();
    });

    it("control overrides", () => {
      expect(
        mount(
          <ToolbarDefaults
            {...mockProps}
            showRedoControl={{
              icon: "#icon-redo",
              label: "Redo",
              description: "Redo last change",
            }}
          />,
        ),
      ).toMatchSnapshot();
    });
  });

  it("#entityTypes", () => {
    expect(
      mount(
        <ToolbarDefaults
          {...mockProps}
          entityTypes={[
            { type: "TEST_1", label: "Test 1" },
            { type: "TEST_2", label: "Test 2" },
            { type: "TEST_3", label: "Test 3" },
          ]}
        />,
      ),
    ).toMatchSnapshot();
  });

  it("#blockTypes", () => {
    expect(
      mount(
        <ToolbarDefaults
          {...mockProps}
          blockTypes={[
            { type: "TEST_1", label: "Test 1" },
            { type: "TEST_2", label: "Test 2" },
            { type: "TEST_3", label: "Test 3" },
          ]}
        />,
      ),
    ).toMatchSnapshot();
  });

  it("#inlineStyles", () => {
    expect(
      mount(
        <ToolbarDefaults
          {...mockProps}
          inlineStyles={[
            { type: "TEST_1", label: "Test 1" },
            { type: "TEST_2", label: "Test 2" },
            { type: "TEST_3", label: "Test 3" },
          ]}
        />,
      ),
    ).toMatchSnapshot();
  });

  it("button label default", () => {
    const wrapper = mount(
      <ToolbarDefaults {...mockProps} inlineStyles={[{ type: "BOLD" }]} />,
    );
    expect(wrapper.find("ToolbarButton").prop("label")).toBe("B");
  });

  it("button label with icon", () => {
    const wrapper = mount(
      <ToolbarDefaults
        {...mockProps}
        inlineStyles={[{ type: "BOLD", icon: "#icon-bold" }]}
      />,
    );
    expect(wrapper.find("ToolbarButton").prop("label")).toBe(null);
  });

  it("button label overrides", () => {
    const wrapper = mount(
      <ToolbarDefaults
        {...mockProps}
        inlineStyles={[
          {
            type: "BOLD",
            label: "Format as bold",
          },
        ]}
      />,
    );
    expect(wrapper.find("ToolbarButton").prop("label")).toBe("Format as bold");
  });

  it("button titles with shortcut", () => {
    expect(
      mount(
        <ToolbarDefaults
          {...mockProps}
          inlineStyles={[{ type: "BOLD", description: null }]}
        />,
      )
        .find("ToolbarButton")
        .prop("title"),
    ).toBe("Ctrl + B");
  });

  it("button titles with shortcut and description", () => {
    expect(
      mount(
        <ToolbarDefaults
          {...mockProps}
          inlineStyles={[{ type: "BOLD", description: "Bold" }]}
        />,
      )
        .find("ToolbarButton")
        .prop("title"),
    ).toBe("Bold\nCtrl + B");
  });
});
