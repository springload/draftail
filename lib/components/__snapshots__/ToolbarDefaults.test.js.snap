// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarDefaults #blockTypes 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={
    Array [
      Object {
        "label": "Test 1",
        "type": "TEST_1",
      },
      Object {
        "label": "Test 2",
        "type": "TEST_2",
      },
      Object {
        "label": "Test 3",
        "type": "TEST_3",
      },
    ]
  }
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_1"
        label="Test 1"
        name="TEST_1"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_1"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 1
          </span>
        </button>
      </ToolbarButton>
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_2"
        label="Test 2"
        name="TEST_2"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_2"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 2
          </span>
        </button>
      </ToolbarButton>
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_3"
        label="Test 3"
        name="TEST_3"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_3"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 3
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #enableHorizontalRule control overrides 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={
    Object {
      "description": "Horizontal rule",
      "icon": "#icon-hr",
      "label": "HR",
    }
  }
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon="#icon-hr"
        label="HR"
        name="HORIZONTAL_RULE"
        onClick={[Function]}
        title="Horizontal rule
- - -"
      >
        <button
          aria-label="Horizontal rule
- - -"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="HORIZONTAL_RULE"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <Icon
            className={null}
            icon="#icon-hr"
            title={null}
          >
            <svg
              aria-hidden={true}
              aria-label={null}
              className="Draftail-Icon "
              height="16"
              role={null}
              viewBox="0 0 1024 1024"
              width="16"
            >
              <use
                xlinkHref="#icon-hr"
              />
            </svg>
          </Icon>
          <span
            className="Draftail-ToolbarButton__label"
          >
            HR
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #enableHorizontalRule works 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={true}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        label="―"
        name="HORIZONTAL_RULE"
        onClick={[Function]}
        title="Horizontal line
- - -"
      >
        <button
          aria-label="Horizontal line
- - -"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="HORIZONTAL_RULE"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            ―
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #enableLineBreak control overrides 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={
    Object {
      "description": "Soft line break",
      "icon": "#icon-br",
      "label": "BR",
    }
  }
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon="#icon-br"
        label="BR"
        name="BR"
        onClick={[Function]}
        title="Soft line break
⇧ + ↵"
      >
        <button
          aria-label="Soft line break
⇧ + ↵"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="BR"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <Icon
            className={null}
            icon="#icon-br"
            title={null}
          >
            <svg
              aria-hidden={true}
              aria-label={null}
              className="Draftail-Icon "
              height="16"
              role={null}
              viewBox="0 0 1024 1024"
              width="16"
            >
              <use
                xlinkHref="#icon-br"
              />
            </svg>
          </Icon>
          <span
            className="Draftail-ToolbarButton__label"
          >
            BR
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #enableLineBreak works 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={true}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        label="↵"
        name="BR"
        onClick={[Function]}
        title="Line break
⇧ + ↵"
      >
        <button
          aria-label="Line break
⇧ + ↵"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="BR"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            ↵
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #entityTypes 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={
    Array [
      Object {
        "label": "Test 1",
        "type": "TEST_1",
      },
      Object {
        "label": "Test 2",
        "type": "TEST_2",
      },
      Object {
        "label": "Test 3",
        "type": "TEST_3",
      },
    ]
  }
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_1"
        label="Test 1"
        name="TEST_1"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_1"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 1
          </span>
        </button>
      </ToolbarButton>
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_2"
        label="Test 2"
        name="TEST_2"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_2"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 2
          </span>
        </button>
      </ToolbarButton>
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_3"
        label="Test 3"
        name="TEST_3"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_3"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 3
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #inlineStyles 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={
    Array [
      Object {
        "label": "Test 1",
        "type": "TEST_1",
      },
      Object {
        "label": "Test 2",
        "type": "TEST_2",
      },
      Object {
        "label": "Test 3",
        "type": "TEST_3",
      },
    ]
  }
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_1"
        label="Test 1"
        name="TEST_1"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_1"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 1
          </span>
        </button>
      </ToolbarButton>
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_2"
        label="Test 2"
        name="TEST_2"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_2"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 2
          </span>
        </button>
      </ToolbarButton>
      <ToolbarButton
        active={false}
        icon={null}
        key="TEST_3"
        label="Test 3"
        name="TEST_3"
        onClick={[Function]}
        title={null}
      >
        <button
          aria-label={null}
          className="Draftail-ToolbarButton"
          data-draftail-balloon={null}
          name="TEST_3"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            Test 3
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  />
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #showRedoControl control overrides 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={
    Object {
      "description": "Redo last change",
      "icon": "#icon-redo",
      "label": "Redo",
    }
  }
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon="#icon-redo"
        label="Redo"
        name="redo"
        onClick={[Function]}
        title="Redo last change
Ctrl + ⇧ + Z"
      >
        <button
          aria-label="Redo last change
Ctrl + ⇧ + Z"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="redo"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <Icon
            className={null}
            icon="#icon-redo"
            title={null}
          >
            <svg
              aria-hidden={true}
              aria-label={null}
              className="Draftail-Icon "
              height="16"
              role={null}
              viewBox="0 0 1024 1024"
              width="16"
            >
              <use
                xlinkHref="#icon-redo"
              />
            </svg>
          </Icon>
          <span
            className="Draftail-ToolbarButton__label"
          >
            Redo
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #showRedoControl works 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={true}
  showUndoControl={false}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        label="↻"
        name="redo"
        onClick={[Function]}
        title="Redo
Ctrl + ⇧ + Z"
      >
        <button
          aria-label="Redo
Ctrl + ⇧ + Z"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="redo"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            ↻
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #showUndoControl control overrides 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={
    Object {
      "description": "Undo last change",
      "icon": "#icon-undo",
      "label": "Undo",
    }
  }
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon="#icon-undo"
        label="Undo"
        name="undo"
        onClick={[Function]}
        title="Undo last change
Ctrl + Z"
      >
        <button
          aria-label="Undo last change
Ctrl + Z"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="undo"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <Icon
            className={null}
            icon="#icon-undo"
            title={null}
          >
            <svg
              aria-hidden={true}
              aria-label={null}
              className="Draftail-Icon "
              height="16"
              role={null}
              viewBox="0 0 1024 1024"
              width="16"
            >
              <use
                xlinkHref="#icon-undo"
              />
            </svg>
          </Icon>
          <span
            className="Draftail-ToolbarButton__label"
          >
            Undo
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
</ToolbarDefaults>
`;

exports[`ToolbarDefaults #showUndoControl works 1`] = `
<ToolbarDefaults
  addBR={[Function]}
  addHR={[Function]}
  blockTypes={Array []}
  currentBlock="unstyled"
  currentStyles={Immutable.OrderedSet []}
  enableHorizontalRule={false}
  enableLineBreak={false}
  entityTypes={Array []}
  inlineStyles={Array []}
  onRequestSource={[Function]}
  onUndoRedo={[Function]}
  showRedoControl={false}
  showUndoControl={true}
  toggleBlockType={[Function]}
  toggleInlineStyle={[Function]}
>
  <ToolbarGroup
    key="styles"
  />
  <ToolbarGroup
    key="blocks"
  />
  <ToolbarGroup
    key="hr-br"
  />
  <ToolbarGroup
    key="entities"
  />
  <ToolbarGroup
    key="undo-redo"
  >
    <div
      className="Draftail-ToolbarGroup"
    >
      <ToolbarButton
        active={false}
        icon={null}
        label="↺"
        name="undo"
        onClick={[Function]}
        title="Undo
Ctrl + Z"
      >
        <button
          aria-label="Undo
Ctrl + Z"
          className="Draftail-ToolbarButton"
          data-draftail-balloon={true}
          name="undo"
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="Draftail-ToolbarButton__label"
          >
            ↺
          </span>
        </button>
      </ToolbarButton>
    </div>
  </ToolbarGroup>
</ToolbarDefaults>
`;

exports[`ToolbarDefaults empty 1`] = `
Array [
  <ToolbarGroup
    key="styles"
  />,
  <ToolbarGroup
    key="blocks"
  />,
  <ToolbarGroup
    key="hr-br"
  />,
  <ToolbarGroup
    key="entities"
  />,
  <ToolbarGroup
    key="undo-redo"
  />,
]
`;
